libdrgn/program.c

```diff
@@ -1489,6 +1522,86 @@ drgn_thread_object(struct drgn_thread *thread, const struct drgn_object **ret)
 	return NULL;
 }
 
+static struct drgn_error *
+drgn_thread_name_linux_kernel(struct drgn_thread *thread, char **ret)
+{
+	struct drgn_error *err;
+	DRGN_OBJECT(comm, drgn_object_program(&thread->object));
+	err = drgn_object_member_dereference(&comm, &thread->object, "comm");
+	if (!err)
+		err = drgn_object_read_c_string(&comm, ret);
+	if (err)
+		return err;
+	return NULL;
+}
+
+static struct drgn_error *
+drgn_thread_name_userspace_live(struct drgn_thread *thread, char **ret)
+{
+#define FORMAT "/proc/%" PRIu32 "/comm"
+	char path[sizeof(FORMAT)
+		- sizeof("%" PRIu32)
+		+ max_decimal_length(uint32_t)
+		+ 1];
+	snprintf(path, sizeof(path), FORMAT, thread->tid);
+#undef FORMAT
+	FILE *file = fopen(path, "r");
+	if (!file)
+		return drgn_error_create_os("fopen", errno, path);
+
+	char buf[TASK_COMM_LEN];
+	size_t bytes_read = 0;
+	while (bytes_read < sizeof(buf)) {
+		size_t bytes_read_iter = fread(buf + bytes_read, 1, sizeof(buf) - bytes_read, file);
+		if (bytes_read_iter == 0) {
+			break;
+		}
+		bytes_read += bytes_read_iter;
+	}
+	fclose(file);
+
+	if (bytes_read > 0) {
+		// Remove trailing new line
+		if (buf[bytes_read - 1] == '\n') {
+			buf[bytes_read - 1] = '\0';
+		}
+		*ret = strndup(buf, sizeof(buf));
+		if (!*ret)
+			return &drgn_enomem;
+	}
+	return NULL;
+}
+
+static struct drgn_error *
+drgn_thread_name_userspace_core(struct drgn_thread *thread, char **ret)
+{
+	struct drgn_error *err = drgn_program_cache_core_dump_notes(thread->prog);
+	if (err)
+		return err;
+	// Core dumps only contain the main thread name so check if this is the main thread.
+	// Otherwise, set ret to NULL which will return None in Python.
+	bool is_main_thread = thread->prog->main_thread && thread->prog->main_thread->tid == thread->tid;
+	if (is_main_thread && thread->prog->core_dump_fname_cached) {
+		*ret = strndup(thread->prog->core_dump_fname_cached, TASK_COMM_LEN);
```

Comment:
If we check that the string is null-terminated when we cache it like I suggested, then this can be `strdup()` instead of `strndup()`.